Fix quick retry immediately after completion marking score as failed

bdach opened the pull request:
Closes https://github.com/ppy/osu/issues/25247.

Likely broke with https://github.com/ppy/osu/pull/25097 (although I haven't really bothered to 100% confirm this).

The scenario involved here is as follows:

1. User completes beatmap by hitting all notes.
2. `checkScoreCompleted()` determines completion, and calls `progressToResults(withDelay: true)`.
3. `progressToResults()` schedules `resultsDisplayDelegate`, which includes a call to `prepareAndImportScoreAsync()`, a second in the future.
4. User presses quick retry hotkey. This calls `Player.Restart(quickRestart: true)`, which invokes `Player.RestartRequested`, which in turn calls `PlayerLoader.restartRequested(true)`, which in turn causes `PlayerLoader` to make itself current, which means that `Player.OnExiting()` will get called.
5. `Player.OnExiting()` sees that `prepareScoreForDisplayTask` is null (because `prepareAndImportScoreAsync()` - which sets it - is scheduled to happen in the future), and as such assumes that the score did not complete. Thus, it marks the score as failed.
6. `Player.Restart()` after invoking `RestartRequested` calls `PerformExit(false)`, which then will unconditionally call `prepareAndImportScoreAsync()`. But the score has already been marked as failed.

The flow above can be described as "convoluted", but I'm not sure I have it in me right now to try and refactor it again. Therefore, to fix, switch the `prepareScoreForDisplayTask` null check in `Player.OnExiting()` to check `GameplayState.HasPassed` instead, as it is not susceptible to the same out-of-order read issue.

bdach committed:
Add failing test coverage for quick retry after completion not changing rank

bdach committed:
Fix quick retry immediately after completion marking score as failed

Closes https://github.com/ppy/osu/issues/25247.

The scenario involved here is as follows:

1. User completes beatmap by hitting all notes.
2. `checkScoreCompleted()` determines completion, and calls
   `progressToResults(withDelay: true)`.
3. `progressToResults()` schedules `resultsDisplayDelegate`, which
   includes a call to `prepareAndImportScoreAsync()`, a second in the
   future.
4. User presses quick retry hotkey.
   This calls `Player.Restart(quickRestart: true)`,
   which invokes `Player.RestartRequested`,
   which in turn calls `PlayerLoader.restartRequested(true)`,
   which in turn causes `PlayerLoader` to make itself current,
   which means that `Player.OnExiting()` will get called.
5. `Player.OnExiting()` sees that `prepareScoreForDisplayTask` is null
   (because `prepareAndImportScoreAsync()` - which sets it -
   is scheduled to happen in the future), and as such assumes that
   the score did not complete. Thus, it marks the score as failed.
6. `Player.Restart()` after invoking `RestartRequested` calls
   `PerformExit(false)`, which then will unconditionally call
   `prepareAndImportScoreAsync()`. But the score has already been marked
   as failed.

The flow above can be described as "convoluted", but I'm not sure I have
it in me right now to try and refactor it again. Therefore, to fix,
switch the `prepareScoreForDisplayTask` null check in
`Player.OnExiting()` to check `GameplayState.HasPassed` instead, as it
is not susceptible to the same out-of-order read issue.

bdach committed:
Add extra test coverage for marking score as failed

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -1115,8 +1115,7 @@ public override bool OnExiting(ScreenExitEvent e)
 
-                // if arriving here and the results screen preparation task hasn't run, it's safe to say the user has not completed the beatmap.
-                if (prepareScoreForDisplayTask == null && DrawableRuleset.ReplayScore == null)
+                if (!GameplayState.HasPassed && DrawableRuleset.ReplayScore == null)
```
Seems correct, but I'd propose a slight simplification:

```diff
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 18d0a65d7a..a1ec0b3167 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -1110,12 +1110,12 @@ public override bool OnExiting(ScreenExitEvent e)
             failAnimationContainer?.Stop();
             PauseOverlay?.StopAllSamples();
 
-            if (LoadedBeatmapSuccessfully)
+            if (LoadedBeatmapSuccessfully && !GameplayState.HasPassed)
             {
-                if (!GameplayState.HasPassed && !GameplayState.HasFailed)
+                if (!GameplayState.HasFailed)
                     GameplayState.HasQuit = true;
 
-                if (!GameplayState.HasPassed && DrawableRuleset.ReplayScore == null)
+                if (DrawableRuleset.ReplayScore == null)
                     ScoreProcessor.FailScore(Score.ScoreInfo);
             }
 

```

bdach replied:
Sure, applied.

I also threw in some extra assertions in https://github.com/ppy/osu/pull/25264/commits/dc7f5cd6edc94275a56b3b5cdb5766c76c84181c for added peace of mind. Not sure if that's what you were imagining when this was brought up in stream chat.

bdach replied:
Turns out I went a bit overboard with the assertion; fixed in https://github.com/ppy/osu/pull/25264/commits/35f30d6135b29b73b68d861bb68564cc951f1a27.

bdach committed:
Share `!HasPassed` condition

bdach committed:
Add preventive assertions concerning submission flow state

bdach committed:
Scale back debug assertion

The import preparation task can actually be non-null when exiting even
if the player hasn't passed:

- fail beatmap
- click import button to import the failed replay
